klippy/extras/pause_resume.py

```diff
@@ -0,0 +1,58 @@
+# Pause/Resume functionality with optional position capturing
+#
+# Copyright (C) 2019  Eric Callahan <arksine.code@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+class PauseResume:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.gcode = self.printer.lookup_object('gcode')
+        self.enable_capture = config.getboolean('capture_default', False)
+        self.recover_velocity = config.getfloat('recover_velocity', 50.)
+        self.captured_position = None
+        self.captured_speed = 0.
+        self.toolhead = self.v_sd = None
+        self.printer.register_event_handler("klippy:ready", self.handle_ready)
+        self.gcode.register_command("PAUSE", self.cmd_PAUSE)
+        self.gcode.register_command("RESUME", self.cmd_RESUME)
+    def handle_ready(self):
+        self.toolhead = self.printer.lookup_object('toolhead')
+        self.v_sd = self.printer.lookup_object('virtual_sdcard', None)
+    def cmd_PAUSE(self, params):
+        capture = (self.gcode.get_str(
+            'CAPTURE', params, str(self.enable_capture)).upper() == "TRUE")
+        if self.v_sd is not None and self.v_sd.is_active():
+            # Printing from virtual sd, run pause command
+            self.v_sd.cmd_M25({})
+        else:
+            self.gcode.respond_info("action:pause")
+        self.toolhead.wait_moves()
+        if capture:
+            self.captured_position = self.toolhead.get_position()
+            self.captured_speed = self.gcode.speed
```

Comment:
A module should not peek into another module's member variables.  Simplest would probably be to add "speed" to gcode.get_status() and use that to get the info.